torch/onnx/_internal/torchscript.py

```diff
@@ -0,0 +1,423 @@
+"""Utilities for manipulating the torch.Graph object and the torchscript."""
+
+# TODO(justinchuby): Move more of the symbolic helper functions here and expose
+# them to the user.
+
+import dataclasses
+import numbers
+import re
+from typing import Any, Dict, Iterable, Iterator, List, Tuple, Union
+
+from typing_extensions import Protocol
+
+import torch
+from torch import _C
+from torch._C import _onnx as _C_onnx
+from torch.onnx import _deprecation
+from torch.onnx._globals import GLOBALS
+from torch.onnx._internal import _beartype
+
+_ATTR_PATTERN = re.compile("^(.+)_(([ifstgz])|(ty))$")
+
+
+class GraphLike(Protocol):
+    """Implements all methods defined in torch.Graph, as well as `op()` and `at()`."""
+
+    def op(
+        self,
+        opname: str,
+        *raw_args: Union[torch.Tensor, _C.Value],
+        outputs: int = 1,
+        **kwargs,
+    ) -> Union[_C.Value, Tuple[_C.Value, ...]]:
+        ...
+
+    def at(
+        self, operator: str, *args, overload_name: str = "", **kwargs
+    ) -> Union[_C.Value, Tuple[_C.Value, ...]]:
+        ...
+
+    def inputs(self) -> List[_C.Value]:
+        ...
+
+    def outputs(self) -> List[_C.Value]:
```

Comment:
I wonder if these are needed for creating this object.

Reply:
IIUC, this is just an Interface class which implementation is calling the implementor method with same name. It mimics `torch.Graph`

I agree it is a bit confusing because the docstring mentions that it "implements" all methods from `torch.Graph`, and then it just declares them.

Maybe the docstring could be reworked to let explicit the intent of serving as abstract class

Reply:
Updated the docstring